proposals/interface-types/CanonicalABI.md

```diff
@@ -0,0 +1,1205 @@
+    }
+
+    flags<$fields>(val) => {
+      values = []
+      for chunk in $fields.chunks(32):
+        bits = 0
+        for i, field in enumerate(chunk):
+          if val[field]:
+            bits |= 1 << i
+        values.push(bits)
+      values
+    }
+
+    record_start<$fields> => {
+      result = []
+      for _ in $fields.len():
+        result.append(lower(direction, dst, gen))
+      assert_eq(gen.next(), record_end<$fields>) # true by construction
+      result
+    }
+
+    # variants are represented with a discriminant plus whatever is the
+    # smallest which fits all possible cases.
+    variant_start<$cases>(label) => {
+      discrim = $cases.index(label)
+      result = [discrim as wasm_i32]
+
+      # Recursively serialize the optional payload
+      if $cases[discrim].has_payload:
+        result.append(lower(direction, dst, gen))
+
+      assert_eq(gen.next(), variant_end<$cases>) # true by construction
+
+      # Upcast each individual type in the payload to the shared type for
+      # each arm of the variant
+      payload_types = flatten_cases(direction, $cases)
+      for i in 0..result.len() - 1:
+        result[i + 1] = upcast(result[i + 1], payload_types[i])
+
+      # If the shared type has more values than this payload generated we
+      # insert zero values to pad
+      for i in result.len() - 1..payload_types.len()
+        result.push(zero_value(payload_types[i]))
+
+      result
+    }
+
+    # Lowering a resource means that we're inserting our handle into a fresh
+    # slot in the destination's table. This insertion operation allocates a new
+    # i32 value which will point to the handle, and note that this also bumps the
+    # handle's internal reference count.
+    handle<$resource>(val) => [TABLES.insert($resource, val, dst)],
+
+    # Buffers are only ever lowered as parameters of an exported function.
+    # Buffers can't be returned from a function so they can never be lowered
+    # for import results.
+    push-buffer<$ty>(val) | pull-buffer<$ty>(val) => {
+      assert(direction == "export")
+      [TABLES.insert_buffer(val, dst)]
+    }
+
+def lower_to_memory(direction, dst, ptr, gen):
+  match gen.next()
+    f32(val) => write_4_bytes(dst, ptr, val),
+    f64(val) => write_8_bytes(dst, ptr, val),
+
+    s8(val) => write_1_byte(dst, ptr, val),
+    u8(val) => write_1_byte(dst, ptr, val),
+    s16(val) => write_2_bytes(dst, ptr, val),
+    u16(val) => write_2_bytes(dst, ptr, val),
+    s32(val) => write_4_bytes(dst, ptr, val),
+    u32(val) => write_4_bytes(dst, ptr, val),
+    s64(val) => write_8_bytes(dst, ptr, val),
+    u64(val) => write_8_bytes(dst, ptr, val),
+
+    char(val) => write_4_bytes(dst, ptr, val),
+
+    string(val) |
+    list<$ty>(val) => {
+      gen.requeue(val) # put `val` back in the generator to get yielded first in `lower`
+      [val_ptr, val_len] = lower(direction, dst, val)
+      write_4_bytes(dst, ptr, val_ptr)
+      write_4_bytes(dst, ptr + 4, val_len)
```

Comment:
Is val_len here the length of the string in USVs, or the length of the string in bytes?

Reply:
This is the length of the string in bytes, and I suspect that some better documentation would be good to make that more clear!

Reply:
To be clear: after adopting a second canonical string encoding, this will need to be changed, with the canonical ABI specifying a `realloc`-based strategy (as described in the first few paras of [this comment](https://github.com/WebAssembly/interface-types/issues/136#issuecomment-861799460)).